x/programs/cmd/simulator/src/lib.rs

```diff
@@ -227,105 +239,70 @@ impl Client {
     /// # Errors
     ///
     /// Returns an error if the serialization or plan fails.
-    pub fn run_plan(&self, plan: &Plan) -> Result<Vec<PlanResponse>, Box<dyn Error>> {
-        run_steps(self.path, plan)
-    }
-
-    /// Performs a single `Execute` step against the simulator and returns the result.
-    /// # Errors
-    ///
-    /// Returns an error if the serialization or single-[Step]-[Plan] fails.
-    pub fn execute_step(&self, key: &str, step: Step) -> Result<PlanResponse, Box<dyn Error>> {
-        let plan = &Plan {
-            caller_key: key.into(),
-            steps: vec![step],
-        };
-
-        run_step(self.path, plan)
+    pub fn run_plan(self, plan: Plan) -> Result<Vec<PlanResponse>, ClientError> {
+        let mut child = Command::new(self.path)
+            .arg("interpreter")
+            .arg("--cleanup")
+            .arg("--log-level")
+            .arg("error")
+            .stdin(Stdio::piped())
+            .stdout(Stdio::piped())
+            .spawn()?;
+
+        let writer = child.stdin.take().ok_or("failed to open stdin").unwrap();
+        let reader = child.stdout.take().ok_or("failed to open stdout").unwrap();
```

Comment:
This is a nit, but it's a good practice to perform the minimal number of mutations necessary. 
```suggestion
        let writer = child.stdin.as_mut().ok_or(ClientError::StdIo)?;
        let reader = child.stdout.as_mut().ok_or(ClientError::StdIo)?;
```
You might be able to come up with a better name for the new error variant though.

Reply:
Could also do something like:
```rust
let Child { stdin, stdout, .. } = Command::new(self.path) // ...
```

Reply:
I like this variant, I'm not feeling very inspired by other error names so `StdIo` will do